optuna/_transform.py

```diff
@@ -0,0 +1,271 @@
+
+    def untransform(self, trans_params: numpy.ndarray) -> Dict[str, Any]:
+        """Untransform a parameter configuration from continuous space to actual values.
+
+        Args:
+            trans_params:
+                A 1-dimensional ``numpy.ndarray`` in the transformed space corresponding to a
+                parameter configuration.
+
+        Returns:
+            A dictionary of an untransformed parameter configuration. Keys are parameter names.
+            Values are untransformed parameter values.
+
+        """
+        assert trans_params.shape == (self._bounds.shape[0],)
+
+        params = {}
+
+        for (name, distribution), encoded_columns in zip(
+            self._search_space.items(), self._column_to_encoded_columns
+        ):
+            trans_param = trans_params[encoded_columns]
+
+            if isinstance(distribution, CategoricalDistribution):
+                # Select the highest rated one-hot encoding.
+                param = distribution.to_external_repr(trans_param.argmax())
+            else:
+                param = _untransform_numerical_param(
+                    trans_param.item(), distribution, self._transform_log
+                )
+
+            params[name] = param
+
+        return params
+
+
+def _transform_search_space(
+    search_space: Dict[str, BaseDistribution], transform_log: bool, transform_step: bool
+) -> Tuple[numpy.ndarray, List[numpy.ndarray]]:
+    assert len(search_space) > 0, "Cannot transform if no distributions are given."
+
+    n_bounds = sum(
+        len(d.choices) if isinstance(d, CategoricalDistribution) else 1
+        for d in search_space.values()
+    )
+
+    bounds = numpy.empty((n_bounds, 2), dtype=numpy.float64)
+    column_to_encoded_columns: List[numpy.ndarray] = []
+
+    bound_idx = 0
+    for distribution in search_space.values():
+        d = distribution
+        if isinstance(d, CategoricalDistribution):
+            n_choices = len(d.choices)
+            bounds[bound_idx : bound_idx + n_choices] = (0, 1)  # Broadcast across all choices.
+            column_to_encoded_columns.append(numpy.arange(bound_idx, bound_idx + n_choices))
+            bound_idx += n_choices
+        elif isinstance(
+            d,
+            (
+                UniformDistribution,
+                LogUniformDistribution,
+                DiscreteUniformDistribution,
+                IntUniformDistribution,
+                IntLogUniformDistribution,
+            ),
+        ):
+            if isinstance(d, UniformDistribution):
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log),
+                    _transform_numerical_param(d.high, d, transform_log),
+                )
+            elif isinstance(d, LogUniformDistribution):
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log),
+                    _transform_numerical_param(d.high, d, transform_log),
+                )
```

Comment:
Now I remembered https://github.com/optuna/optuna/pull/1558. UniformDistribution and LogUniformDistribution only accept `[low, high)`. But some optimizer (e.g. pycma and cmaes) samples the upper bound value, then it will be fallback to the independent sampler.

We have some solutions for this problem.

1. Add a `_SearchSpaceTransform`'s option to cap the upper bound with a tiny value
1. Always cap the upper bound in `_SearchSpaceTransform` (without an option).
1. Cap the upper bound in CmaEsSampler and PyCmaSampler.

I think the second solution is enough for all samplers. But it seems a controversial topic.
May I have your opinion?

Reply:
You're right. I'm wondering if we shouldn't go with 3 in this PR to keep it reasonably scoped. However, I agree that 2 sounds like a fair compromise otherwise. We'd be able to isolate this somewhat controversial logic to a single point.

Reply:
@c-bata what do you think?

Reply:
> You're right. I'm wondering if we shouldn't go with 3 in this PR to keep it reasonably scoped. However, I agree that 2 sounds like a fair compromise otherwise. We'd be able to isolate this somewhat controversial logic to a single point.

Agree. The only concern is that all upper bounds (including categorical, int, and discrete) will be capped if we choose 3. But I guess it does not affect the optimization performance. I made this resolved.

(Sorry, I'm busy today. So I'll review other changes next week.)

Reply:
(Thanks a lot. Don't mean to stress but do you have any estimates on when? If you're busy, we could also invite another reviewer or if your time is limited maybe schedule a mob session to reduce the number of roundtrips.)